src/main/java/ch/njol/skript/effects/EffTeleport.java

```diff
@@ -167,22 +192,56 @@ protected TriggerItem walk(Event e) {
 					}
 				}
 
-				TriggerItem.walk(next, e);
+				TriggerItem.walk(next, event);
 			}
-			Variables.removeLocals(e); // Clean up local vars, we may be exiting now
+			Variables.removeLocals(event); // Clean up local vars, we may be exiting now
 			SkriptTimings.stop(timing);
 		});
 		return null;
 	}
 
 	@Override
-	protected void execute(Event e) {
-		// Nothing needs to happen here, we're executing in walk
+	protected void execute(Event event) {
+		assert false;
 	}
 
 	@Override
-	public String toString(@Nullable Event e, boolean debug) {
-		return "teleport " + entities.toString(e, debug) + " to " + location.toString(e, debug);
+	public String toString(@Nullable Event event, boolean debug) {
+		return "teleport " + entities.toString(event, debug) + " to " + location.toString(event, debug) +
+				teleportFlags == null ? "" : " retaining " + teleportFlags.toString(event, debug);
+	}
+
+	private void teleport(@NotNull Teleport teleport, @Nullable TeleportFlags paperTeleportFlags, @NotNull Entity entity, @NotNull Location location, SkriptTeleportFlag... skriptTeleportFlags) {
+		if (!TELEPORT_FLAGS_SUPPORTED || paperTeleportFlags == null || skriptTeleportFlags == null) {
+			teleport.teleport(entity, location);
+			return;
+		}
+		Stream<TeleportFlag> teleportFlags = Arrays.stream(skriptTeleportFlags).flatMap(teleportFlag -> {
+			if (teleportFlag == SkriptTeleportFlag.RETAIN_DIRECTION)
+				return Stream.of(TeleportFlag.Relative.PITCH, TeleportFlag.Relative.YAW);
+			return Stream.of(teleportFlag.getTeleportFlag());
+		}).filter(Objects::nonNull);
+
+		paperTeleportFlags.teleport(entity, location, teleportFlags.toArray(TeleportFlag[]::new));
+	}
```

Comment:
```suggestion
	private void teleport(@NotNull Entity entity, @NotNull Location location, SkriptTeleportFlag... skriptTeleportFlags) {
		if (location.getWorld() == null) {
			location = location.clone();
			location.setWorld(entity.getWorld());
		}
	
		if (!TELEPORT_FLAGS_SUPPORTED || skriptTeleportFlags == null) {
			entity.teleport(location);
			return;
		}
		Stream<TeleportFlag> teleportFlags = Arrays.stream(skriptTeleportFlags).flatMap(teleportFlag -> {
			if (teleportFlag == SkriptTeleportFlag.RETAIN_DIRECTION)
				return Stream.of(TeleportFlag.Relative.PITCH, TeleportFlag.Relative.YAW);
			return Stream.of(teleportFlag.getTeleportFlag());
		}).filter(Objects::nonNull);

		entity.teleport(location, teleportFlags.toArray(TeleportFlag[]::new));
	}
```
I don't think the functional interface route is doing anything beneficial here. It'd be much simpler to just call the teleport methods directly, rather than have two interfaces and static methods and whatnot.

Reply:
The functional interface is still needed to avoid static imports, paper methods and is being used as a passive enum. Inside the function interface it also collects all the skript based enums into an TeleportFlag array.

I have simplified the function interface to make it easier for those to understand functional interfaces.

The location world null check clone applied.

Reply:
The functional interface still isn't necessary, only the enum is. See https://discord.com/channels/135877399391764480/836220422223036467/1286092678383140967.